arangod/Aql/QueryRegistry.h

```diff
@@ -51,7 +53,9 @@ class QueryRegistry {
   /// With keepLease == true the query will be kept open and it is guaranteed
   /// that the caller can continue to use it exclusively.
   /// This is identical to an atomic sequence of insert();open();
-  TEST_VIRTUAL void insert(QueryId id, Query* query, double ttl, bool isPrepare, bool keepLease);
+  TEST_VIRTUAL void insert(
+    QueryId id, Query* query, double ttl, bool isPrepare, bool keepLease,
+    std::unique_ptr<arangodb::cluster::CallbackGuard>&& = nullptr);
```

Comment:
This looks seriously broken!

Reply:
It's not broken. To give some context. This PR registers a callback in the RebootTracker to remove a query and its transaction if the server that requested it rebooted.

The registered callback in the RebootTracker has to be cleaned up at some point, which is what the CallbackGuard is for: as soon as it is destroyed, it removes the callback from the RebootTracker. Thus it is placed together with the query in the query registry.

As this is not used at all places a query is created, the guard is optional. That is totally fine.